packages/core/src/default-values.ts

```diff
@@ -0,0 +1,439 @@
+
+    for (const elementValue of elementValues) {
+      if (elementValue === undefined) {
+        continue;
+      }
+
+      if (!Array.isArray(elementValue)) {
+        throw new Error('Expected array value for sliced element');
+      }
+
+      const matchingItems: any[] = this.getMatchingSliceValues(elementValue, slice, slicing);
+      sliceValues.push(matchingItems);
+    }
+
+    this.valueStack.push({
+      type: 'slice',
+      path,
+      values: sliceValues,
+    });
+  }
+
+  getMatchingSliceValues(elementValue: any[], slice: SliceDefinitionWithTypes, slicing: VisitorSlicingRules): any[] {
+    const matchingItems: any[] = [];
+    for (const arrayItem of elementValue) {
+      const sliceName: string | undefined =
+        arrayItem[SLICE_NAME_KEY] ?? getValueSliceName(arrayItem, [slice], slicing.discriminator, this.schema.url);
+
+      if (sliceName === slice.name) {
+        matchingItems.push(arrayItem);
+      }
+    }
+
+    // Make sure at least slice.min values exist
+    for (let i = matchingItems.length; i < slice.min; i++) {
+      if (isComplexTypeCode(slice.type[0].code)) {
+        const emptySliceValue = Object.create(null);
+        matchingItems.push(emptySliceValue);
+
+        // push onto input array so that it propagates upwards as well
+        elementValue.push(emptySliceValue);
+      }
+    }
+
+    return matchingItems;
+  }
+
+  onExitSlice(): void {
+    const sliceValuesContext = this.valueStack.pop();
+    if (!sliceValuesContext) {
+      throw new Error('Expected value context to exist in onExitSlice');
+    }
+
+    for (const sliceValueArray of sliceValuesContext.values) {
+      for (let i = sliceValueArray.length - 1; i >= 0; i--) {
+        const sliceValue = sliceValueArray[i];
+        if (SLICE_NAME_KEY in sliceValue) {
+          delete sliceValue[SLICE_NAME_KEY];
+        }
+      }
+    }
+  }
+
+  getDefaultValue(): any {
+    return this.rootValue;
+  }
+}
+
+function applyMinimums(
+  parent: any,
+  key: string,
+  element: InternalSchemaElement,
+  elements: Record<string, InternalSchemaElement>
+): void {
+  const existingValue = getValueAtKey(parent, key, element, elements);
+
+  if (element.min > 0 && existingValue === undefined) {
+    if (isComplexTypeCode(element.type[0].code)) {
+      if (element.isArray) {
+        setValueAtKey(parent, [Object.create(null)], key, element);
+      } else {
+        setValueAtKey(parent, Object.create(null), key, element);
+      }
+    }
+  }
+}
+
+function setValueAtKey(parent: any, value: any, key: string, element: InternalSchemaElement): void {
+  if (key.includes('.')) {
+    throw new Error('key cannot be nested');
+  }
+
+  let resolvedKey = key;
+
+  if (key.includes('[x]')) {
+    const code = element.type[0].code;
+    resolvedKey = key.replace('[x]', capitalize(code));
+  }
+
+  if (value === undefined) {
+    delete parent[resolvedKey];
+  } else {
+    parent[resolvedKey] = value;
+  }
+}
+
+function getValueAtKey(
```

Comment:
We can revisit this in future PR, but this method has a lot of overlap with `getTypedPropertyValue`

Reply:
Yeah, part of the problem (and why I added `getValueAtKey` is that `getTypedPropertyValue` doesn't return the `Object.is` equivalent value in some cases (e.g. `isEmpty(resultValue)` and `Array.isArray(resultValue)` which was important for this code.